### PR TITLE
[Task] Add namespace crud

### PR DESCRIPTION
**Pull request by sufuf3**
I finished Namespace CRUD, please help me review this PR.
Thank you.

### CODE REVIEW

File: src/entity/namespace.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"time"
+
+	"gopkg.in/mgo.v2/bson"
+)
+
+// the const for NamespaceCollectionName
+const (
+	NamespaceCollectionName string = "namespace"
```
Comment (John-Lin) [inline src/entity/namespace.go]: `namespaces`

**Review by John-Lin — CHANGES_REQUESTED** [review]

File: src/entity/namespace.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+// Namespace is the structure for namespace info
+type Namespace struct {
+	ID        bson.ObjectId     `bson:"_id,omitempty" json:"id" validate:"-"`
+	Name      string            `bson:"name" json:"name" validate:"required,k8sname"`
+	Labels    map[string]string `bson:"labels,omitempty" json:"labels" validate:"required,dive,keys,alphanum,endkeys,required,alphanum"`
```
Comment (John-Lin) [inline src/entity/namespace.go]: I think we don't need labels here

**Review by John-Lin — CHANGES_REQUESTED** [review]

File: src/kubernetes/namespaces.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+// CreateNamespace will create the namespace by the namespace object
+func (kc *KubeCtl) CreateNamespace(namespace *corev1.Namespace) (*corev1.Namespace, error) {
+	return kc.Clientset.CoreV1().Namespaces().Create(namespace)
+}
+
+// DeleteNamespace will delete the namespace by the namespace name
+func (kc *KubeCtl) DeleteNamespace(name string) error {
+	deletePolicy := metav1.DeletePropagationBackground
```
Comment (John-Lin) [inline src/kubernetes/namespaces.go]: Do we need DeletePropagationBackground option here?

Comment (sufuf3) [reply src/kubernetes/namespaces.go]: I referred to https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L340-L344
IMO, maybe it's good to need?
What do you think?

Comment (John-Lin) [reply src/kubernetes/namespaces.go]: I think we should use **Foreground** waiting for all resource to be killed. see https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation

```
The object exists in the key-value store until the garbage collector
deletes all the dependents whose ownerReference.blockOwnerDeletion=true
from the key-value store.  API sever will put the "foregroundDeletion"
finalizer on the object, and sets its deletionTimestamp.  This policy is
cascading, i.e., the dependents will be deleted with Foreground.
```

**Review by John-Lin — CHANGES_REQUESTED** [review]

File: src/server/handler_namespace.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	})
+	defer session.Close()
+
+	// Check whether this name has been used
+	n.ID = bson.NewObjectId()
+	n.CreatedAt = timeutils.Now()
+
+	if err := namespace.CreateNamespace(sp, &n); err != nil {
+		if errors.IsAlreadyExists(err) {
+			response.Conflict(req.Request, resp.ResponseWriter, fmt.Errorf("Namespace Name: %s already existed", n.Name))
```
Comment (John-Lin) [inline src/server/handler_namespace.go]: `Namespace: %s already existed`

File: src/server/handler_namespace.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	defer session.Close()
+
+	n := entity.Namespace{}
+	if err := session.FindOne(entity.NamespaceCollectionName, bson.M{"_id": bson.ObjectIdHex(id)}, &n); err != nil {
+		response.BadRequest(req.Request, resp.ResponseWriter, err)
+		return
+	}
+
+	if err := namespace.DeleteNamespace(sp, &n); err != nil {
+		if errors.IsAlreadyExists(err) {
```
Comment (John-Lin) [inline src/server/handler_namespace.go]:
```
if errors.IsNotFound(err) {
```

**Review by John-Lin — CHANGES_REQUESTED** [review]

**Comment by John-Lin** [comment]
@sufuf3 Please add namespace api usage in API.md

**Comment by sufuf3** [comment]
@John-Lin I added namespace api usage in API.md and modified the files.
Please help me review this PR. Thank you.

File: API.md [hunk 6]
```diff
@@ -1361,3 +1366,94 @@ Response Data:
...
+
+Request Data:
+
+```json
+{
+  "name": "awesome",
+}
+```
+
+Response Data:
```
Comment (John-Lin) [inline API.md:1391]: response data should be the namespace entity not create success message

Comment (sufuf3) [reply API.md:1391]: I copied from service, sorry that I forgot to change the response data.
Updated it.

**Review by John-Lin — APPROVED** [review]

File: src/namespace/namespace.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	n := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   namespace.Name,
+		},
+	}
+	_, err := sp.KubeCtl.CreateNamespace(&n)
+	return err
+}
+
+// DeleteNamespace willl delete namespace
```
Comment (WakeupTsai) [inline src/namespace/namespace.go]: 小錯字 will

**Review by WakeupTsai — APPROVED** [review]

File: src/namespace/namespace.go (hunk 7 above)
Comment (John-Lin) [reply src/namespace/namespace.go]: 我修了

**Pull request merged**